framework/aster-frame/src/task/scheduler.rs

```diff
@@ -1,64 +1,146 @@
 // SPDX-License-Identifier: MPL-2.0
 
+use crate::config::SCHED_DEBUG_LOG;
 use crate::prelude::*;
-use crate::sync::SpinLock;
-use crate::task::Task;
+use crate::sync::{SpinLock, SpinLockGuard};
+use crate::task::{NeedResched, ReadPriority, Task};
 
-use lazy_static::lazy_static;
+use log::debug;
 
-lazy_static! {
-    pub(crate) static ref GLOBAL_SCHEDULER: SpinLock<GlobalScheduler> =
-        SpinLock::new(GlobalScheduler { scheduler: None });
+pub(crate) static GLOBAL_SCHEDULER: spin::Once<SpinLock<GlobalScheduler>> = spin::Once::new();
+
+pub fn init() {
+    GLOBAL_SCHEDULER.call_once(|| SpinLock::new(GlobalScheduler { scheduler: None }));
 }
 
 /// A scheduler for tasks.
 ///
-/// An implementation of scheduler can attach scheduler-related information
-/// with the `TypeMap` returned from `task.data()`.
-pub trait Scheduler: Sync + Send {
-    fn enqueue(&self, task: Arc<Task>);
+/// Operations on the scheduler should be performed with interrupts disabled,
+/// which has been ensured by the callers of the `GLOBAL_SCHEDULER`.
+/// Therefore, implementations of this trait do not need to worry about interrupt safety.
+pub trait Scheduler<T: NeedResched + ReadPriority = Task>: Sync + Send {
+    /// Called when a task enters a runnable state.
+    /// Add the task to the scheduler.
+    fn enqueue(&self, task: Arc<T>);
+
+    /// Called when the task is no longer alive.
+    /// Remove the task-related from the scheduler.
+    ///
+    /// A similar method in the linux scheduler is `dequeue_task`.
+    ///
+    /// Return `true` if the task was in the scheduler.
+    fn remove(&self, task: &Arc<T>) -> bool;
```

Comment:
Why we need a `remove` method? The `remove` is called when a thread exits. But at that time, the thread should not be in the scheduler queue, right?

Based on the current implementation, the `pick_next_task` will remove the task from scheduler, so the ruuning task should not be in scheduler queue.

Reply:
✅ for the current scheduler which preserves no task-related statistic. For O(1) scheudler, the method is needed to purge unnecessary sched entities from the scheduler.

I was thinking if the func name, `remove`, and its semantics are proper. Or, use the followings?

```rust
pub trait Scheduler<T: NeedResched + ReadPriority = Task>: Sync + Send {
    /// Called when a task enters a runnable state.
    /// Add the task to the scheduler.
    fn enqueue(&self, task: Arc<T>);

    /// Drop the task from the scheduler if it is in the scheduler.
    /// Return `true` if the task was in the scheduler.
    fn dequeue(&self, task: &Arc<T>) -> bool;

    // FIXME: named as purge/remove_dead?                ******
    /// Called when the task is no longer alive.
    /// Remove all the task-related from the scheduler.
    /// Return `true` if the task was in the scheduler.
    fn remove(&self, task: &Arc<T>) -> bool {
        self.dequeue(task) && {
            self.clear(task);
            true
        }
    }

    /// Clear the all states of the given task from the scheduler.
    fn clear(&self, task: &Arc<T>) {}

    /// Choose the most appropriate task eligible to run next.
    fn pick_next_task(&self) -> Option<Arc<T>>;
    // ...
}
```

And impl for the current scheduler:

```rust
impl Scheduler for PreemptiveRRScheduler {
    fn dequeue(&self, task: &Arc<T>) -> bool {/* dequeue the task from the queues and return true */}
    fn remove(&self, task: &Arc<T>) -> bool {
        // do nothing
        false
    }
    fn clear(&self, task: &Arc<T>) {/* still empty here */}
}
```

Besides, can `remove_task` and `remove` from `Scheduler` trait be used to handle killed threads?

Reply:
Ok, I think we can keep the `remove` method. Since the real usage is in O(1) scheduler, we can rethink the interface at that time.

> Besides, can `remove_task` and `remove` from `Scheduler` trait be used to handle killed threads?

I think what you referred to as "killing a thread" means that a thread is terminated without invoking `Task::exit`. In the current kernel implementation, such a situation doesn't actually exist. Killing a thread (e.g., triggering an exception or receiving a signal like SIGTERM) only leads to the main loop in `usr_task_entry`(in aster-std/src/thread/task.rs) exiting, but eventually, it still invokes the `Task::exit` method. So, I'm not quite clear why there would be residual scheduling information for dead threads in the scheduler. However, I don't oppose temporarily keeping this interface, if your O(1) scheduler does need it.